nebula/ui/components/inAppAuth/VPNInAppAuthenticationInputs.qml

```diff
@@ -46,24 +46,61 @@ ColumnLayout {
 
         spacing: VPNTheme.theme.listSpacing
 
-        VPNTextField {
-            id: textInput
-            objectName: base.objectName + "-textInput"
+        RowLayout {
+            id: textInputSection
             Layout.fillWidth: true
-            _placeholderText: _inputPlaceholderText
-            Keys.onReturnPressed: col.submitInfo(textInput)
-            onDisplayTextChanged: if (hasError) hasError = false
+            spacing: VPNTheme.theme.windowMargin / 2
+
+            VPNTextField {
+                id: textInput
+                objectName: base.objectName + "-textInput"
+                Layout.fillWidth: true
+                _placeholderText: _inputPlaceholderText
+                Keys.onReturnPressed: col.submitInfo(textInput)
+                onDisplayTextChanged: if (hasError) hasError = false
+            }
+
+            VPNPasteButton {
+                id: textInputPasteButton
+                objectName: base.objectName + "-textInputPasteButton"
+                Layout.preferredWidth: VPNTheme.theme.rowHeight
+                Layout.preferredHeight: VPNTheme.theme.rowHeight
+                height: undefined
+                width: undefined
+                onClicked: {
+                    textInput.paste();
+                }
+            }
         }
 
-        VPNPasswordInput {
-            id: passwordInput
-            objectName: base.objectName + "-passwordInput"
-            _placeholderText: _inputPlaceholderText
+        RowLayout {
+            id: passwordInputSection
             Layout.fillWidth: true
-            Keys.onReturnPressed: col.submitInfo(passwordInput)
-            onTextChanged: if (hasError) hasError = false
+            spacing: VPNTheme.theme.windowMargin / 2
+
+            VPNPasswordInput {
+                id: passwordInput
+                objectName: base.objectName + "-passwordInput"
+                _placeholderText: _inputPlaceholderText
+                Keys.onReturnPressed: col.submitInfo(passwordInput)
+                Layout.fillWidth: true
+                onTextChanged: if (hasError) hasError = false
+            }
+
+            VPNPasteButton {
+                id: passwordInputPasteButton
+                objectName: base.objectName + "-passwordInputPasteButton"
+                Layout.preferredWidth: VPNTheme.theme.rowHeight
+                Layout.preferredHeight: VPNTheme.theme.rowHeight
+                height: undefined
+                width: undefined
+                onClicked: {
+                    passwordInput.paste();
```

Comment:
What about if we move the onclick in the button impl passing the corresponding object?

Reply:
done. way better.